plugins/security/nexus-ldap-plugin/src/main/java/org/sonatype/security/realms/ldap/internal/persist/DefaultLdapConfigurationManager.java

```diff
@@ -50,135 +50,120 @@
-    this.configurationValidator = checkNotNull(configurationValidator);
-    this.eventBus = eventBus;
+    this.validator = checkNotNull(validator);
+    this.eventBus = checkNotNull(eventBus);
     this.securitySystem = checkNotNull(securitySystem);
+    this.cache = Maps.newLinkedHashMap();
+    this.cachePrimed = false;
   }
 
   @Override
   public synchronized void clearCache() {
-    ldapConfiguration = null;
-    // fire event
+    cache.clear();
+    cachePrimed = false;
     eventBus.post(new LdapClearCacheEvent(this));
   }
 
   @Override
   public synchronized void setServerOrder(final List<String> orderdServerIds)
-      throws InvalidConfigurationException
+      throws IllegalArgumentException
   {
-    final List<CLdapServerConfiguration> ldapServers = getConfiguration().getServers();
-    final ValidationResponse vr = configurationValidator.validateLdapServerOrder(ldapServers, orderdServerIds);
-    if (vr.getValidationErrors().size() > 0) {
-      throw new InvalidConfigurationException(vr);
-    }
-
-    // build a map so its easier
-    final Map<String, CLdapServerConfiguration> idToServerMap = Maps.newHashMap();
-    for (CLdapServerConfiguration ldapServer : ldapServers) {
-      idToServerMap.put(ldapServer.getId(), ldapServer);
-    }
-    // now reorder them
-    final List<CLdapServerConfiguration> newOrderedldapServers = Lists.newArrayList();
-    for (String serverId : orderdServerIds) {
-      newOrderedldapServers.add(idToServerMap.get(serverId));
+    checkNotNull(orderdServerIds);
+    final LinkedHashMap<String, LdapConfiguration> configuration = getConfiguration();
+    final Set<String> newOrder = Sets.newHashSet(orderdServerIds);
+    final Set<String> configIds = configuration.keySet();
+    checkArgument(newOrder.equals(configIds), "Bad ordering: %s (existing IDs %s)", newOrder, configIds);
+
+    for (LdapConfiguration config : configuration.values()) {
+      config.setOrder(orderdServerIds.indexOf(config.getId()));
+      configurationSource.update(config);
     }
-    getConfiguration().setServers(newOrderedldapServers);
-    save();
+    clearCache();
   }
 
   @Override
-  public synchronized List<CLdapServerConfiguration> listLdapServerConfigurations() {
-    return ImmutableList.copyOf(getConfiguration().getServers());
+  public synchronized List<LdapConfiguration> listLdapServerConfigurations() {
+    return ImmutableList.copyOf(getConfiguration().values());
   }
 
   @Override
-  public synchronized CLdapServerConfiguration getLdapServerConfiguration(final String id)
-      throws InvalidConfigurationException,
-             LdapServerNotFoundException
+  public synchronized LdapConfiguration getLdapServerConfiguration(final String id)
+      throws LdapServerNotFoundException
   {
-    for (CLdapServerConfiguration ldapServer : getConfiguration().getServers()) {
-      if (ldapServer.getId().equals(id)) {
-        return ldapServer;
-      }
+    checkNotNull(id);
```

Comment:
Is a non-null id here sufficient or should it also be non-empty to be considered valid?